### ISSUE TITLE
collapsing a chain of gates

### ISSUE THREAD

**Issue by ajavadia** [Qiskit/qiskit-terra#5775]
There are a few compiler passes right now related to peephole optimization: 
- Collect2qBlocks
- ConsolidateBlocks
- Optimize1qGates
- Optimize1qDecomposition
- https://github.com/Qiskit/qiskit-terra/pull/4747

We should unify these into one general pass which is able to walk the circuit and collect+consolidate these gates. This should rely on the `DAGDependency` graph otherwise it will miss some opportunities when gates commute but are not next to each other. This should also accept a filter function that tells the pass to collect those gates that meet a certain condition (I believe retworkx can already do this). The filter functions could be:
- gate.num_qubits == 1
- gate.num_qubits <= 2
- gate.is_clifford
- etc.

Also I think we should just combine the collection phase with the consolidation phase. We can call it `CollapseChains`. I haven't found a use case for collecting a chain of gates but not collapsing them into an equivalent unitary.

**Comment by ajavadia** [comment]
So the quaternion business in `Optimize1qGate` can be gotten rid of I think if we just collapse into a unitary and use the OneQubitEulerDecomposer on it. In which case I think it becomes the same as `Optimize1qDecomposition`. Why are these 2 different again?

I guess in the case of cliffords I see the point that we want to collapse it into an N-qubit Clifford rather than a N-qubit Unitary.

**Comment by mtreinish** [comment]
> So the quaternion business in `Optimize1qGate` can be gotten rid of I think if we just collapse into a unitary and use the OneQubitEulerDecomposer on it. In which case I think it becomes the same as `Optimize1qDecomposition`. Why are these 2 different again?
> 

There are 2 reasons. The first is that the quaternion pass will simplify to more pulse efficient gates U3 -> U2 or U1 which the decomposer pass doesn't do yet (this is in progress in #5554 ). But, the other aspect is runtime performance, after #5554 the results between the 2 passes will be identical for the old ibmq basis but the quaternion pass is still faster. Even with the tuning we did in #5682 the quaternion pass is about 40-50% faster than the decomposer pass for the u1, u2, u3, id, cx basis (I haven't tested the performance impact with #5554 yet, but it will make the decomposer pass slightly slower because it will construct 2 unitaries to compare the results between the 2 basis sets in the decomposer). While that's not major, and #5554 will change the preset pass managers to always use the decomposer pass, it's still enough that if you're working with the old basis having the faster pass around as an option seems worthwhile.

**Comment by levbishop** [comment]
I like this direction but I have a few comments:

1. Collecting in general is not unique and there may be different heuristics you'd like for choosing a good collection. For a trivial example, collecting nq<=2 has ambiguity of which block gets the X in:
   ```
    ----+--------
        | 
     ---+- X -+--
              |
    ----------+--
    ```
    Mostly this doesn't matter for things like TwoQubitDecomposer since we can do but for collecting nq<=3 then there could be very different ways of collecting the gates, with very different decompositions. The pass in #4747 has a heuristic to try to collect the smallest number of blocks but a) this is just a heuristic; and b) smallest number of blocks is not necessarily one's goal. The algorithm might get a bit hairy to cover all needs, but I guess all the more reason to separate it all out into one common place to put a spotlight on it.
2. I agree with @ajavadia that dropping the quaternions in favor of collecting the unitary *should* be just as good, and I'm a bit surprised that the quaternions are faster. I should do some benchmarking but is the difference in the collection phase or the synthesis phase? If its in the collection phase  there's around 10x of overhead in the Operator wrapping of the numpy matrices for these chains of 2x2 matrices, that we can very easily work around.
```python
In [269]: ops = [qiskit.quantum_info.random.random_unitary(2) for _ in range(1000)]                                                                       

# current implementation, more or less
In [270]: %timeit functools.reduce(qiskit.quantum_info.Operator.compose, ops)                                                                                  
13.8 ms ± 784 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

# directly access the numpy arrays: 10x speedup
In [271]: %timeit qiskit.quantum_info.Operator(functools.reduce(np.dot, (o._data for o in reversed(ops))))                                                          
1.46 ms ± 32.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

# log-depth vectorized multiplication tree gets another 30% speedup (for long-enough chains)
def mat_reduce(m): 
    while len(m) % 2 == 0: 
    m = np.matmul(m[::2], m[1::2]) 
    return functools.reduce(np.dot, m) 

In [272]: %timeit qiskit.quantum_info.Operator(mat_reduce([o._data for o in reversed(ops)]))                                                              
1.07 ms ± 15 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

3. I vote for @mtreinish suggestion to have both `filter_function()` and `collapse_function()` which allows for things like collapsing cliffords but also for collapsing non-unitary circuit elements such as a string of reset instructions. We can provide an implementation of collapse function to gather to unitary since that is the common case.
4. Maybe a convenient interface for `collapse_function()` should be `self, inout_sub_dag -> None` so that it gets called once for each collected chain, passed to it as a sub-dag that it gets to modify inplace?

**Comment by mtreinish** [comment]
>     2\. I agree with @ajavadia that dropping the quaternions in favor of collecting the unitary _should_ be just as good, and I'm a bit surprised that the quaternions are faster. I should do some benchmarking but is the difference in the collection phase or the synthesis phase? If its in the collection phase  there's around 10x of overhead in the Operator wrapping of the numpy matrices for these chains of 2x2 matrices, that we can very easily work around.

Here is a profile of the `Optimize1qGatesDecomposition`:

![Screenshot_2021-02-11_16-27-09](https://user-images.githubusercontent.com/2447371/107700808-062d9480-6c86-11eb-9e85-1438772f29a5.png)

generated with:

```python3
from qiskit.transpiler.passes import BasisTranslator
from qiskit.transpiler.passes import Optimize1qGatesDecomposition
from qiskit.circuit.random import random_circuit
from qiskit.converters import circuit_to_dag
from qiskit.test.mock.backends import FakeManhattan
from qiskit.circuit.equivalence_library import SessionEquivalenceLibrary as sel
import time
import cProfile

backend_real = FakeManhattan()
real_basis = backend_real.configuration().basis_gates

input_circuit = random_circuit(65, 4096, measure=True, seed=42)
dag = circuit_to_dag(input_circuit)
bt_real = BasisTranslator(sel, real_basis)
basis_dag = bt_real.run(dag)
optimize_pass = Optimize1qGatesDecomposition(real_basis)
start = time.time()
cProfile.run("result = optimize_pass.run(basis_dag)", filename='/tmp/operator.prof')
stop = time.time()
print("Optimize 1Q decomposition %s" % str(stop - start))
```

It looks like it's spending most of the time in the decomposer. We can tune the operator constructor further, and probably removing the compose and just doing the dot product manually will speed things up more. But there is a diminishing margin there because `Operator ` construction is only 1/2 as much time cumulatively as the actual decomposer calls. But we can open another issue and/or PRs to work on this separately. 

Like I said before I agree we only need the decomposition pass by default moving forward (and #5554 makes that change). I just thought because the quaternion pass is still faster to produce the same result we shouldn't remove (well deprecate) it yet while it's still faster. That way people can still use it if they're working in a u1, u2, u3 basis.

**Comment by mtreinish** [comment]
As a first step on this issue I was going to add a collect_runs method to `DAGDependency`that just called retworkx's collect runs method. But I think we're going to need a new different method to collect chains on a `DAGDependency`. The retworkx function `collect_runs` is about finding linear paths of nodes inside a graph where there is only successor node (it's basically just a porting of the old dagcircuit method: https://github.com/Qiskit/retworkx/blob/master/src/lib.rs#L1102-L1171 ), but the structure in `DAGDependency` doesn't match that expectation for finding chains in that way. For example with this circuit:

```
        ┌───┐     ┌──────────┐┌──────────┐┌──────────┐ ░ ┌─┐      
  q0_0: ┤ H ├──■──┤ U1(3.14) ├┤ U1(3.14) ├┤ U1(3.14) ├─░─┤M├──────
        └───┘┌─┴─┐└──┬───┬───┘└──────────┘└──────────┘ ░ └╥┘┌─┐   
  q0_1: ─────┤ X ├───┤ X ├─────────■───────────────────░──╫─┤M├───
             └───┘   └─┬─┘       ┌─┴─┐       ┌───┐     ░  ║ └╥┘┌─┐
  q0_2: ───────────────■─────────┤ X ├───────┤ H ├─────░──╫──╫─┤M├
                                 └───┘       └───┘     ░  ║  ║ └╥┘
meas: 3/══════════════════════════════════════════════════╩══╩══╩═
                                                          0  1  2 
```

The `DAGDependency` graph looks like:

![dep_dag](https://user-images.githubusercontent.com/2447371/108532532-f0335b80-72a5-11eb-9967-71e6c0fdb8b4.png)

If we wanted to collect a chain of 1q gates there retworkx's `collect_runs()` function won't work because it can't find the chain of 3 u1s, instead it would basically put each 1q gate in a separate run.

**Comment by 1ucian0** [comment]
> There are a few compiler passes right now related to peephole optimization:

Add this one: `CXCancellation`

**Comment by ajavadia** [comment]
@mtreinish right. So I think we need to modify the function a bit for DAGDependency. We are interested in collecting the maximal block of gates where all gates meet a certain criteria (here `gate.num_qubits < 1`). So for each node, I think we need to look at its children as well as siblings (i.e. those that share the same parents). And keep going down the graph until we cannot grow the block anymore. So in your example H plus 3 U1s will be collected. 

But also your example shows how this graph is better than the DAGCircuit, because the DAGCircuit won't allow us to bundle H with the U1s (they are separated by a CNOT, but the U1s actually can commute before the CNOT).

**Comment by ajavadia** [comment]
I'm also good with the suggestion above for the structure of the base pass, and providing an implementation for the common case where we want to collect and collapse into a Unitary.

**Comment by ajavadia** [comment]
Ok this is actually simpler to implement, because code already exists for it in Qiskit. 

Finding the longest sequence of gates acting on some N qubits, considering commutativity, is described in Section 5.4.2 of this paper: https://arxiv.org/pdf/1909.05270.pdf

So we need to adapt a special case of this[ template matching algorithm in Qiskit](https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/transpiler/passes/optimization/template_matching/template_matching.py) where it considers every gate that meets the collection criteria to be a match (instead of looking at a template library which it does now).

**Comment by mtreinish** [comment]
Looking through the paper and the template matching code I'm thinking there is a performance tradeoff here we're going to need to keep in mind. While I can look at implementing the `PatternMatch` algorithm natively in retworkx (unless someone else would like to give it a try), which will speed it up to some degree (how much so I'm not sure because we'll still need to callback to python for anything beyond the graph structure) and we'll definitely want to do longer term, I still think there will be a tradeoff here. When I look at the complexity of the `PatternMatch` algorithm vs what we do today especially for the level1 preset pass managers I think will make it too slow. I think it would be a good fit for level 3 and maybe even level 2.

So I'm starting to think we're not going to want to base all of our peephole optimization passes around dag dependency from a performance perspective. For example, the tradeoff between faster (especially after my PR series, which starts in https://github.com/Qiskit/qiskit-terra/pull/5915 ) but less thorough optimization in `Optimize1qGatesDecomposition` which is using the `DAGCircuit`'s run collection feels like a good fit for level 1, but passes based on an abstract chain collapsing pass which does optimization using `DAGDependency` and `PatternMatch` would be a better fit for higher optimization levels and will provide better optimization at the cost of runtime.

**Comment by ajavadia** [comment]
I think as a first step we should just try to make a pass for block collection that is based on the PatternMatch, which already exists in Qiskit. Once we know it works well and it can find all blocks, then we can profile performance and tune it. I agree we may have to do simple passes at lower optimization levels.

**Comment by alexanderivrii** [comment]
I would be willing to give this a try.

**Comment by alexanderivrii** [comment]
Here are a few thoughts on the _block collection_ part of this; I will open a draft PR later today.

I agree with @levbishop's [comment](https://github.com/Qiskit/qiskit-terra/issues/5775#issuecomment-777523637) that collecting is generally not unique and that we should concentrate on lightweight heuristic approaches. So a reasonably simple and useful strategy seems something like this: "starting from the inputs (leafs) of DagDependency, choose heuristically largest blocks of nodes matching certain criteria". Note that when a node is (marked as) collected, some of its immediate successors can now become leafs and may also be eligible for collecting (depending on the collection criteria). 

Let's associate to each node its `in_degree`, that is the number of its direct predecessors; a node is an input if and only if its `in_degree` is 0.

Below I will describe several different collection strategies.

1. As per @ajavadia's [original suggestion](https://github.com/Qiskit/qiskit-terra/issues/5775#issue-799687491), a large class of collecting methods consists of _iteratively_ collecting blocks of nodes that match a certain filtering function `filter_fn`. Examples of this include collecting blocks of swap gates, blocks of linear gates (CXs and SWAPs), blocks of Clifford gates, blocks of single-qubit gates, blocks of two-qubit gates, etc.. Here, "iteratively" means that once a node is collected, the `in_degree` of its immediate successors is decreased by 1, and some of these successors may be also collectible into the same current block. 

Taking collecting Clifford gates as an example, the overall strategy to process a DagDependency could be to repeat the following until no more uncollected nodes remain: 
- iteratively collect a maximal block of non-Clifford gates (= iteratively collect nodes with `in_degree = 0` that do not match `filter_fn`)
- iteratively collect a maximal block of Clifford gates (= iteratively collect nodes with  `in_degree = 0` that match `filter_fn`)

Intuitively, extracting larger blocks of non-Clifford gates helps finding larger blocks of Clifford gates later on.

2. A somewhat different collecting method consists of collecting blocks of _commuting_ nodes that match a certain filtering function, as for example grouping commuting 2-qubit gates in #8034. This amounts to simply taking all nodes that have `in_degree=0` that also match the given criteria (e.g., it's a 2-qubit gate, or specifically a CZ-gate, etc.). Indeed, by the canonicity of DagDependency, every pair of input nodes commutes, and once a node is chosen to be a part of a commuting block, none of its parents can be. So this is _not_ an iterative process. An overall strategy to process a DagDependency for #8034 could be to repeat the following until no more uncollected nodes remain: 
- iteratively collect nodes with `in_degree = 0` that do not match `filter_fn` (the method [1])
- repeat (until nothing gets collected)
  - collect nodes with `in_degree = 0` that do match `filter_fn` into a block of commuting nodes  
 
Note that in general there can be multiple consecutive blocks of commuting gates, so the second step above is repeated multiple times until nothing gets collected.

3. One can also tie in @kdk's suggestion [suggestion](https://github.com/Qiskit/qiskit-terra/issues/7387#issuecomment-1017638771) for #7387 as a method to collect parallel blocks of `in_degree=0` nodes, where `parallel` means that all nodes in the same block must have pairwise disjoint qubits. In a DagDependency, an edge between two nodes means that the associated gates do not commute and hence must have intersecting sets of qubits, thus this is once again not an iterative process, but there is a "packing" problems of choosing a maximal non-overlapping set of "in_degree=0" nodes. Possibly a greedy algorithm would work well here. Thus, an overall strategy to process a DagDependency for depth minimization would be to repeat the following until no more uncollected nodes remain: 
- collect non-overlapping `in_degree=0` nodes into a block.

I believe one can show that the depth of the resulting circuit is necessarily equal to the total number of collected blocks.

Any comments and additional thoughts?

I also didn't quite figure out how to tie the above with collecting runs of single-qubit gates, or two-qubit gates, or multi-qubit gates. Well, for single-qubit gates, possibly the method [1] for collecting blocks of single-qubit gates, and then splitting these blocks by individual qubits, could work. But I don't see a good solution for collecting two-qubit (or multi-qubit) gates.

**Comment by alexanderivrii** [comment]
I am contemplating to change the API for `collapse_function` (proposed in https://github.com/Qiskit/qiskit-terra/issues/5775#issuecomment-772019684 and temporarily implemented in #8319) from 

```
@abstractmethod
def collapse_function(self, runs, dag):
    pass
```
to
```
@abstractmethod
def collapse_function(self, single_run):
    pass
```
This would simplify the pass that collects blocks of CX and SWAP gates and replaces each block by a `LinearFunction`: the `collapse_function` method would only need to construct a `LinearFunction` from a block of linear gates, and not need to worry about iterating over all blocks or about calling `dag.replace_block_with_op` -- these would now be handled by `CollapseChainsPass.run()`. This would similarly simplify other similar passes, such as collecting blocks Clifford gates and replacing them by a `Clifford`,  collecting blocks of SWAP gates and replacing them by a `Permutation`. 

However, with this change, the `collapse_function` will not have a global view of the DAG, but I cannot think of a use-case where this would be helpful.

What do you think? Better yet, are there any strong objections to changing/simplifying the API as described?

**Comment by ajavadia** [comment]
I think this is fine. The use cases of collapsing I can think of don't really require a global view of the DAG. Just needs to know what to collapse into (Clifford, LinearFunction, Unitary, etc.). And some options such as the ones here, which tell whether to collapse any given chain, or not, depending on the cost trade-offs of whether collapsing would be beneficial down the road: 
https://github.com/Qiskit/qiskit-terra/blob/d59f2386cacf1e822b8791ff20c1e22bda36b83d/qiskit/transpiler/passes/optimization/consolidate_blocks.py#L40

**Comment by ShellyGarion** [comment]
@alexanderivrii - Can this issue be closed now since #8907 was merged?

**Comment by alexanderivrii** [comment]
@ShellyGarion, not quite, #8907 does two things: (1) it provides a very general transpiler pass for collecting and collapsing blocks of nodes, which in theory should be able to fit all of different block collection strategies discussed here, and (2) it provides a specific implementation of the `collect` function that combines blocks of nodes that match a given filter function. Thus (2) covers use-cases such as "collect blocks of SWAP gates", "collect blocks of Clifford gates", "collect blocks consisting of 2-qubit gates only", but it does not cover other use-cases such as "collect blocks with each block consisting of single-qubit and two-qubit gates over the same pair of qubits" or "collect blocks of commuting gates". So I feel that to fully close this issue we need to provide `collect` functions that cover these cases as well.
